peekingduck/pipeline/nodes/input/visual.py

```diff
@@ -0,0 +1,259 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Reads inputs from multiple visual sources
+- image or video file on local storage
+- folder of images or videos
+- online cloud source
+- CCTV or webcam live feed
+"""
+
+from pathlib import Path
+from typing import Any, Dict, Optional, Union
+
+from peekingduck.pipeline.nodes.input.utils.preprocess import resize_image
+from peekingduck.pipeline.nodes.input.utils.read import VideoNoThread, VideoThread
+from peekingduck.pipeline.nodes.node import AbstractNode
+
+
+class Node(AbstractNode):  # pylint: disable=too-many-instance-attributes
+    """Receives visual sources as inputs.
+
+    Inputs:
+        |none_input_data|
+
+    Outputs:
+        |img_data|
+
+        |filename_data|
+
+        |pipeline_end_data|
+
+        |saved_video_fps_data|
+
+    Configs:
+        filename (:obj:`str`): **default = "video.mp4"**. |br|
+            Filename of the MP4 file if media is exported. |br|
+            If source is a directory of files, then filename is the current file being
+            processed.
+        frames_log_freq (:obj:`int`): **default = 100**. |br|
+            Logs frequency of frames passed in CLI
+        mirror_image (:obj:`bool`): **default = False**. |br|
+            Flag to set extracted image frame as mirror image of input stream.
+        resize (:obj:`Dict`):
+            **default = { do_resizing: False, width: 1280, height: 720 }** |br|
+            Dimension of extracted image frame.
+        saved_video_fps (:obj:`int`): **default = 10**. |br|
+            FPS to be used to output the MP4 file after livestream is processed and
+            exported.  FPS is dependent on running machine performance.
+        source (:obj:`int`):
+            **default = https://storage.googleapis.com/peekingduck/videos/wave.mp4**. |br|
+            Input source can be: |br|
+            - filename : local image or video file |br|
+            - directory name : all media files will be processed |br|
+            - http URL for online cloud source : http[s]://... |br|
+            - rtsp URL for CCTV : rtsp://... |br|
+            - 0 for webcam live feed |br|
+            Refer to `OpenCV documentation
+            <https://docs.opencv.org/4.5.5/d8/dfe/classcv_1_1VideoCapture.html>`_
+            for more technical information.
+        threading (:obj:`bool`): **default = False**. |br|
+            Flag to enable threading when reading frames from camera.
+            The FPS can increase up to 30%.
```

Comment:
I cannot remember exactly, but I believe we turned this to False by default because it could not exit gracefully in Linux if True? Should we add a word of caution here after verifying this again?

Reply:
`threading` is False by default because of potential gotcha's.
If True by default and input is a file, it will read the entire file faster than can be processed and end prematurely, if user forgets to enable `buffer_frames`.
But if `threading` and `buffer_frames` are both True by default, then for live stream/webcam inputs, it will likely produce a laggy video, especially if the system is slower than source.